### PR TITLE
Manifest file generator

### PR DESCRIPTION
**Pull request by ffoulkes**
Utility to generate manifest files from a json master file.

Can create both `deps.cmake` and `version.json` from a single source of truth.

The input manifest can be overridden, allowing us to have alternate configurations. With a few tweaks to the cmake files, we can eliminate the `custom-grpc.sh` and `recent-grpc.sh` script files, the `RECENT_PKGS` cmake option, and the conditional logic in the `deps.cmake` file, all of which were done to facilitate experimental development with different versions of the packages.

I'm debating whether this belongs under the `scripts` directory or a new TLD for administrative files (e.g. `admin`).

We also need to consider criteria for inclusion in the version file. I'm currently selecting based on whether the package has a defined `tag`.  The content of the version file is an open matter, too.

I'm also toying with the idea of implementing a "query" mode that would accept a package name and return a string containing one or more values for the specified package. CMake could use this instead of `deps.cmake`.

It gets even better starting with CMake 3.19, which adds support for reading and interpreting JSON files in CMakeLists.txt. 😎

Thoughts?

### CODE REVIEW

**Comment by 5abeel** [comment]
Good progress towards maintaining a single source of truth that is easily human readable. This is where I was thinking of taking the version.json file I created a draft on in #33. Will close that PR now

One recommendation would be to move scrips/manifest.json to main folder version.json -- so that it is easily locatable for a user on a main GitHub page as a quick reference. The Python script can still remain inside a scripts folder

File: scripts/manifest/manifest.json [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        "sha": "827c769e5fc98e0f2a34c47cef953cc6328abced",
+        "url": "https://github.com/gflags/gflags.git"
+    },
+    "glog": {
+        "name": "glog",
+        "sha": "a8e0007e96ff96145022c488e367da10f835c75d",
+        "tag": "v0.6.0-rc1",
+        "url": "https://github.com/google/glog.git"
+    },
+    "grpc": {
```
Comment (5abeel) [inline scripts/manifest/manifest.json]: Defaults have been updated with the merge of #41 Please upate version, SHA and tag

File: scripts/manifest/manifest.json [hunk 2]
```diff
@@ -0,0 +1,63 @@
+{
+    "abseil": {
```
Comment (5abeel) [inline scripts/manifest/manifest.json]: Defaults have been updated with the merge of #41 Please upate

File: scripts/manifest/manifest.json [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        "tag": "v1.11.0",
+        "url": "https://github.com/google/googletest.git"
+    },
+    "json": {
+        "label": "tags/v3.10.5^2~8",
+        "name": "json",
+        "sha": "760304635dc74a5bf77903ad92446a6febb85acf",
+        "url": "https://github.com/nlohmann/json.git"
+    },
+    "protobuf": {
```
Comment (5abeel) [inline scripts/manifest/manifest.json]: Defaults have been updated with the merge of #41 Please upate

**Comment by ffoulkes** [comment]
> Good progress towards maintaining a single source of truth that is easily human readable. This is where I was thinking of taking the version.json file I created a draft on in #33. Will close that PR now
> 
> One recommendation would be to move scrips/manifest.json to main folder version.json -- so that it is easily locatable for a user on a main GitHub page as a quick reference. The Python script can still remain inside a scripts folder

Replacing `versions.json` with this file would remove the need for the `--f version` option, which would remove the need for the `--format` altogether (for now, at least).

I agree with you about the placement of the files. I hid my json file to leave pride of place to yours. 

`version.json` isn't quite right as the filename, though.
- _version_ is singular. There's more than one version in the file.
- The file isn't about versions, it's about components.

If you don't like `manifest.json`, how about `components.json`?

**Comment by ffoulkes** [comment]
Ooh. I just got an idea.

### Proposal

Add a `defines` property whose value is a dictionary of key-value pairs. The Python script turns these into `set` commands in the generated cmake file. The build script uses the defined variables, rather than `RECENT_PKGS`, to implement conditional behavior.

### components.json

```json
    "protobuf": {
        "name": "protobuf",
        "sha": "6b5d8db01fe47478e8d400f550e797e6230d464e",
        "tag": "v25.0",
        "url": "https://github.com/google/protobuf.git",
        "version": "25.0",
        "defines": {
            "ABSL_PROVIDER": "package",
            "BUILD_SHARED_LIBS": "off"
        }
    },
```

### deps.cmake

```cmake
set(PROTOBUF_GIT_URL "https://github.com/google/protobuf.git")
set(PROTOBUF_GIT_TAG "6b5d8db01fe47478e8d400f550e797e6230d464e") # v25.0
set(PROTOBUF_VERSION "25.0")
set(PROTOBUF_ABSL_PROVIDER "package")
set(PROTOBUF_BUILD_SHARED_LIBS "off")
```

### protobuf.cmake

```cmake
# In older versions of protobuf, the primary CMakeLists.txt file
# is in the cmake subdirectory.
if(DEFINED PROTOBUF_SOURCE_SUBDIR)
  set(_source_subdir SOURCE_SUBDIR ${PROTOBUF_SOURCE_SUBDIR})
endif()

# Protobuf v23.x generates unresolved external references to
# ThreadSafeArena::ThreadCache _thread_cache if BUILD_SHARED_LIBS=ON.
if(DEFINED PROTOBUF_BUILD_SHARED_LIBS)
  set(_build_shared_libs -DBUILD_SHARED_LIBS=${PROTOBUF_BUILD_SHARED_LIBS})
endif()

# Protobuf v23.x bundles its own copy of Abseil.
# We suppress it in favor of our own version or the
# version bundles with gRPC.
if(DEFINED PROTOBUF_ABSL_PROVIDER)
  set(_absl_provider -DPROTOBUF_ABSL_PROVIDER=${PROTOBUF_ABSL_PROVIDER})
endif()
```

### Notes

For this version of the manifest, the BUILD_SHARED_LIBS and ABSL_PROVIDER options are in effect. The SOURCE_SUBDIR option is not.

Any time we hit a new changed condition, we define a variable for it, add code+comment to support it in the cmake files, and specify it in the manifest file if it applies to the version of the component we're picking.

When the problem is safely in the rearview mirror, we make change permanent and remove the definition from the manifest file.

**Comment by ffoulkes** [comment]
Possible additions to the component properties:

- `patch` - name of the patch file to apply
- `license` - the package's license type
- `license_file` - name or path to license file (I think we're going to need to add a `licenses` directory, in which case _name_ will be sufficient.)

**Comment by ffoulkes** [comment]
Update:

- Renamed manifest to `components.json` and moved it to TLD.
- Renamed script to `components.py` and moved it to `scripts/`.
- Removed support for `--format` parameter.
- Updated `abseil`, `grpc`, and `protobuf` to new versions.

With the changes to the script, you can apply component updates by issuing the command:

```bash
./scripts/components.py -o cmake/deps.cmake
```

The change will take effect the next time you configure cmake.

**Comment by ffoulkes** [comment]
I thought of a new output: a Markdown file containing a table that lists the components, together with their version numbers and URLs.

**Comment by ffoulkes** [comment]
Updates:
- Implemented the `defines` property.

**Comment by 5abeel** [comment]
This looks good. I am happy with taking this approach

**Review by 5abeel — APPROVED** [review]
LGTM

**Pull request merged**